File: 0707-design-linked-list/0707-design-linked-list.py
class Node:
    def __init__(self, val=None, next=None):
        self.val = val
        self.next = next


class MyLinkedList:

    def __init__(self):
        self.head = None
        self.size = 0

    def get(self, index: int) -> int:
        if index < 0 or index >= self.size:
            return -1

        count = 0
        cur = self.head
        while cur:
            if count == index:
                return cur.val
            count += 1
            cur = cur.next

    def addAtHead(self, val: int) -> None:
        node = Node(val, self.head)
        self.head = node
        self.size += 1

    def addAtTail(self, val: int) -> None:
        if self.head is None:
            self.head = Node(val)
            self.size += 1
            return

        cur = self.head
        while cur.next:
            cur = cur.next
        cur.next = Node(val)
        self.size += 1

    def addAtIndex(self, index: int, val: int) -> None:
        if index < 0 or index > self.size:
            return

        if index == 0:
            self.addAtHead(val)
            return

        count = 0
        cur = self.head
        while cur:
            if count == index - 1:
                node = Node(val, cur.next)
                cur.next = node
                self.size += 1
                return
            count += 1
            cur = cur.next

    def deleteAtIndex(self, index: int) -> None:
        if index < 0 or index >= self.size:
            return

        if index == 0:
            self.head = self.head.next
            self.size -= 1
            return

        count = 0
        cur = self.head
        while cur:
            if count == index - 1:
                cur.next = cur.next.next
                self.size -= 1
                return
            count += 1
            cur = cur.next

        


# Your MyLinkedList object will be instantiated and called as such:
# obj = MyLinkedList()
# param_1 = obj.get(index)
# obj.addAtHead(val)
# obj.addAtTail(val)
# obj.addAtIndex(index,val)
# obj.deleteAtIndex(index)

File: 0707-design-linked-list/0707-design-linked-list.java
class MyLinkedList {
    class Node{
        int data;
        Node next;
        Node(int data){
            this.data = data;
        }
    }
    private Node head;
    private int size;

    public MyLinkedList() {
        this.head = null;
        this.size = 0;
        
    }
    
    public int get(int index) {
        if(index < 0 || index >= size){
            return -1;
        }
        Node cur = head;
        for(int i = 0; i < index; i++){
            cur = cur.next;
        }
        
        return cur.data;

    }
    
    public void addAtHead(int val) {
        Node newnode = new Node(val);
        newnode.next=head;
        head = newnode;
        size++;

    }
    
    public void addAtTail(int val) {
        
        if(head == null){
            addAtHead(val);
        }else{
            Node newnode = new Node(val);
            Node cur = head;

            while(cur.next != null){
                cur = cur.next;
            }
            cur.next = newnode;
            size++;
        }
    }
    
    public void addAtIndex(int index, int val) {
        if(index < 0 || index > size){
            return;
        }
        if(index == 0){
            addAtHead(val);
            return;
        }

        Node cur = head;
        for(int i = 0; i < index - 1; i++){
            cur = cur.next;
        }
        Node newnode = new Node(val);
        newnode.next = cur.next;
        cur.next = newnode;
        size++;
    }
    
    public void deleteAtIndex(int index) {
        if(index < 0 || index >= size){
            return;
        }
        if(index == 0){
            head = head.next;
            size--;
            return;
        }
        Node cur = head;
        for(int i = 0; i < index - 1; i++){
            cur = cur.next;
        }
        Node temp = cur.next;
        cur.next = temp.next;
        size--;

    }
}

/**
 * Your MyLinkedList object will be instantiated and called as such:
 * MyLinkedList obj = new MyLinkedList();
 * int param_1 = obj.get(index);
 * obj.addAtHead(val);
 * obj.addAtTail(val);
 * obj.addAtIndex(index,val);
 * obj.deleteAtIndex(index);
 */